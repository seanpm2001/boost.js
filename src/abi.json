[
  "function boosts(uint256) view returns (string strategyURI, address token, uint256 balance, address guard, uint256 start, uint256 end, address owner)",
  "function claimTokens(tuple(uint256 boostId, address recipient, uint256 amount) claim, bytes signature)",
  "function claimed(address, uint256) view returns (bool)",
  "function createBoost(tuple(string strategyURI, address token, uint256 balance, address guard, uint256 start, uint256 end, address owner) boost)",
  "function depositTokens(uint256 boostId, uint256 amount)",
  "function eip712ClaimStructHash() view returns (bytes32)",
  "function nextBoostId() view returns (uint256)",
  "function withdrawRemainingTokens(uint256 boostId, address to)"
]
