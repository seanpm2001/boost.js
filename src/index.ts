import fetch from 'cross-fetch';
import { JsonRpcProvider, Web3Provider } from '@ethersproject/providers';
import { Contract } from '@ethersproject/contracts';
import abi from './abi.json';

export const IPFS_GATEWAY = 'pineapple.fyi';
export const BOOST_ADDRESS = '0xaf8b6af86044821eED74E49057De62fB5C48e061';

interface Boost {
  strategyURI: string;
  token: string;
  balance: string;
  guard: string;
  start: number;
  end: number;
  owner: string;
}

interface Strategy {
  strategy: string;
  params: Record<string, any>;
}

export async function getBoost(boostId: number, chainId: number): Promise<Boost> {
  const provider = new JsonRpcProvider(`https://rpc.brovider.xyz/${chainId.toString()}`);
  const contract = new Contract(BOOST_ADDRESS, abi, provider);

  const boost = await contract.boosts(boostId);

  return {
    strategyURI: boost[0],
    token: boost[1],
    balance: boost[2].toString(),
    guard: boost[3],
    start: boost[4].toNumber(),
    end: boost[5].toNumber(),
    owner: boost[6]
  };
}

export async function getStrategy(
  strategyURI: string,
  gateway: string = IPFS_GATEWAY
): Promise<Strategy> {
  const url = strategyURI
    .replace('ipfs://', `https://${gateway}/ipfs/`)
    .replace('ipns://', `https://${gateway}/ipns/`);

  return fetch(url).then((res) => res.json());
}

export async function getNextBoostId(chainId: number): Promise<number> {
  const provider = new JsonRpcProvider(`https://rpc.brovider.xyz/${chainId.toString()}`);
  const contract = new Contract(BOOST_ADDRESS, abi, provider);

  const nextBoostId = await contract.nextBoostId();

  return nextBoostId.toNumber();
}

export async function createBoost(
  web3: Web3Provider,
  strategyURI: string,
  token: string,
  balance: string,
  guard: string,
  start: number,
  end: number,
  owner: string
): Promise<any> {
  const signer = web3.getSigner();
  const contract = new Contract(BOOST_ADDRESS, abi, signer);
  const params = {
    strategyURI,
    token,
    balance,
    guard,
    start,
    end,
    owner
  };

  return await contract.createBoost(params);
}

export async function claimTokens(
  web3: Web3Provider,
  boostId: number,
  recipient: string,
  amount: string,
  signature: string
): Promise<any> {
  const signer = web3.getSigner();
  const contract = new Contract(BOOST_ADDRESS, abi, signer);
  const claim = {
    boostId,
    recipient,
    amount
  };

  return await contract.claimTokens(claim, signature);
}
