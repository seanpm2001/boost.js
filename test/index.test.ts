import { getBoost, getNextBoostId, getStrategy } from '../src';

const chainId = 5;

describe('Boost', () => {
  it('getBoost', async () => {
    const boost = await getBoost(4, chainId);
    expect(boost).toMatchSnapshot();
  });

  it('getStrategy', async () => {
    const strategyURI = 'ipfs://bafkreihfwpcbmgq7llxnwbwekql2kjwc33g2a6pqj5ylfxasbcnappage4';
    const strategy = await getStrategy(strategyURI);
    expect(strategy).toMatchSnapshot();
  });

  it('getNextBoostId', async () => {
    const nextBoostId = await getNextBoostId(chainId);
    expect(nextBoostId).toEqual(expect.any(Number));
  });
});
