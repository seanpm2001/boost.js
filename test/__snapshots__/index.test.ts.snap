// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boost getBoost 1`] = `
Object {
  "balance": "100000000000000000",
  "end": 1668341642,
  "guard": "0xF63EB3f569C6cB8F5Cf37caD183790Ed1b251c91",
  "owner": "0xeF8305E140ac520225DAf050e2f71d5fBcC543e7",
  "start": 1668082442,
  "strategyURI": "ipfs://bafkreihfwpcbmgq7llxnwbwekql2kjwc33g2a6pqj5ylfxasbcnappage4",
  "token": "0xdc31Ee1784292379Fbb2964b3B9C4124D8F89C60",
}
`;

exports[`Boost getStrategy 1`] = `
Object {
  "params": Object {
    "amount": "10000000000000000",
    "proposal": "0x1b48a83c44e323275a605b244a05bde89918fb9ec86be7bb83792eb26e544441",
    "type": "fixed",
  },
  "strategy": "snapshot",
}
`;
